### PR TITLE
Pages Routes Dev

### PR DESCRIPTION
**Pull request by CarmenPopoviciu**

### CODE REVIEW

File: packages/wrangler/src/pages/functions/routes-validation.ts [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+			throw new FatalError(
+				getRoutesValidationErrorMessage(
+					RoutesValidationError.OVERLAPPING_RULES,
+					routesPath
+				),
+				1
+			);
+		}
+	} catch (err) {
+		// TODO what do we want here
```
Comment (jrf0110) [inline packages/wrangler/src/pages/functions/routes-validation.ts:97]: I'd think consumers of `validateRoutes` would handle these errors if necessary. No need to try/catch

File: packages/wrangler/templates/pages-dev-pipeline.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+import worker from "__ENTRY_POINT__";
+// @ts-ignore entry point will get replaced
+export * from "__ENTRY_POINT__";
+
+const transformToRegex = (filter: string) => {
+	return filter.replace("*", ".*");
+};
+
+const routes = {
+	// @ts-ignore routes are injected
```
Comment (jrf0110) [inline packages/wrangler/templates/pages-dev-pipeline.ts:15]: instead of using `@ts-ignore` we can be a bit type-safer and do something like this:

```tsx
declare const __ROUTES__: unknown

if (!isRoutesJSONSpec(__ROUTES__)) {
  throw new Error('Missing global __ROUTES__ value')
}
```

File: packages/wrangler/templates/pages-dev-pipeline.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+	include: __ROUTES__.include.map(transformToRegex),
+	// @ts-ignore routes are injected
+	exclude: __ROUTES__.exclude?.map(transformToRegex) || [],
```
Comment (jrf0110) [inline packages/wrangler/templates/pages-dev-pipeline.ts:18]: `exclude` isn't optional, so why do we need the `?`?

File: packages/wrangler/src/pages/functions/routes-validation.ts (hunk 1 above)
Comment (CarmenPopoviciu) [reply packages/wrangler/src/pages/functions/routes-validation.ts:97]: 👍

File: packages/wrangler/templates/pages-dev-pipeline.ts (hunk 3 above)
Comment (CarmenPopoviciu) [reply packages/wrangler/templates/pages-dev-pipeline.ts:18]: we don't! I think I just forgot it in there. good catch!

File: packages/wrangler/src/pages/dev.tsx [hunk 4]
```diff
@@ -1,17 +1,19 @@
 import { execSync, spawn } from "node:child_process";
-import { existsSync } from "node:fs";
+import { existsSync, readFileSync } from "node:fs";
 import { homedir, tmpdir } from "node:os";
 import { join, resolve } from "node:path";
 import { watch } from "chokidar";
-import { build as workerJsBuild } from "esbuild";
+import { build, build as workerJsBuild } from "esbuild";
```
Comment (GregBrimble) [inline packages/wrangler/src/pages/dev.tsx:6]: Bit of a weird import here. Did you mean to import the same thing twice like this?

File: fixtures/pages-functions-with-routes-app/public/index.html [hunk 5]
```diff
@@ -1,6 +1,8 @@
 <!DOCTYPE html>
 <html>
 	<body>
-		<h1>Bonjour de notre projet ✨pages-functions-with-routes-app✨!</h1>
+		<h1>
+			Bonjour de notre projet &#10024; pages-functions-with-routes-app &#10024;!
```
Comment (rozenmd) [inline fixtures/pages-functions-with-routes-app/public/index.html:5]: Not to be a total nerd but you probably want `Bienvenue sur notre projet` 🙈

File: packages/wrangler/src/pages/dev.tsx (hunk 4 above)
Comment (CarmenPopoviciu) [reply packages/wrangler/src/pages/dev.tsx:6]: ah, so glad you caught this. I saw it as well and was confused as to why it is this way, but didn't fix immediately so here we are are. I think some auto import thing went terribly wrong here. will fix

File: fixtures/pages-functions-with-routes-app/public/index.html (hunk 5 above)
Comment (CarmenPopoviciu) [reply fixtures/pages-functions-with-routes-app/public/index.html:5]: awwww...best review of them all <3

File: packages/wrangler/src/pages/functions/routes-validation.ts [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+/**
+ * Returns true if the given `routingSpec` object contains at least
+ * `MIN_FUNCTIONS_ROUTES_INCLUDE_RULES` include routing rules
+ */
+function hasIncludeRules(routesJSON: RoutesJSONSpec): boolean {
+	// sanity check
+	// this should never be the case, because of the context from which tehse validation fns are
+	// called, but let's not assume anything
+	if (!routesJSON || !routesJSON.include) {
+		throw new Error();
```
Comment (GregBrimble) [inline packages/wrangler/src/pages/functions/routes-validation.ts:111]: Can we add a message to this error (and other similar ones)? Even something simple like "hasIncludeRules was called out of place" would be a good hint to anyone that does hit this.

File: packages/wrangler/src/pages/dev.tsx [hunk 7]
```diff
@@ -307,8 +310,44 @@ export const Handler = async ({
...
+		if (existsSync(routesJSONPath)) {
+			const routesJSONContents = readFileSync(routesJSONPath, "utf-8");
+
+			try {
+				validateRoutes(JSON.parse(routesJSONContents), directory);
+			} catch (err) {
+				throw err;
+			}
+
+			entrypoint = join(tmpdir(), `${Math.random().toString(36).slice(2)}.js`);
```
Comment (GregBrimble) [inline packages/wrangler/src/pages/dev.tsx:327]: I know I wrote this bit, and it's 100% not required for this PR, but this is becoming a frequently used enough pattern, that I'd be tempted to make a function out of this: `makeTemporaryRandomFile(ext: string)`

File: packages/wrangler/src/pages/dev.tsx [hunk 8]
```diff
@@ -307,8 +310,44 @@ export const Handler = async ({
...
+
+			await build({
```
Comment (GregBrimble) [inline packages/wrangler/src/pages/dev.tsx:329]: We should watch `_route.json` and re-build whenever it changes.

Comment (CarmenPopoviciu) [reply packages/wrangler/src/pages/dev.tsx:329]: ✅

File: packages/wrangler/src/pages/functions/routes-validation.ts (hunk 6 above)
Comment (CarmenPopoviciu) [reply packages/wrangler/src/pages/functions/routes-validation.ts:111]: ✅

File: packages/wrangler/src/pages/dev.tsx (hunk 7 above)
Comment (CarmenPopoviciu) [reply packages/wrangler/src/pages/dev.tsx:327]: will make a separate PR out of that, mainly because most of the places we generate these temp random files, we usually specify the file name and randomise just the (let's call it) 'hash' segment. Plus we'd need to support file extensions as well, since we'd be calling this to generate json/esm/js files (from scanning our pages code). So I feel this should have a scope of its own. Sound good?

Comment (GregBrimble) [reply packages/wrangler/src/pages/dev.tsx:327]: Yup, ofc. Like I said, totally not required in this PR. And honestly, not even something you need to do. I flagged it mostly for myself.

File: packages/wrangler/src/pages/dev.tsx (hunk 4 above)
Comment (CarmenPopoviciu) [reply packages/wrangler/src/pages/dev.tsx:6]: ✅

File: fixtures/pages-functions-with-routes-app/public/index.html (hunk 5 above)
Comment (CarmenPopoviciu) [reply fixtures/pages-functions-with-routes-app/public/index.html:5]: ✅

File: packages/wrangler/templates/pages-dev-pipeline.ts (hunk 3 above)
Comment (CarmenPopoviciu) [reply packages/wrangler/templates/pages-dev-pipeline.ts:18]: ✅

File: packages/wrangler/templates/pages-dev-pipeline.ts (hunk 2 above)
Comment (CarmenPopoviciu) [reply packages/wrangler/templates/pages-dev-pipeline.ts:15]: ✅

Comment (CarmenPopoviciu) [reply packages/wrangler/templates/pages-dev-pipeline.ts:11]: @GregBrimble where would we import `isRoutesJSONSpec` from tho? The above fails with 
```
✘ [ERROR] [plugin alias] unrecognized module: ./../src/pages/functions/routes-validation
```
This must be one of those esbuild things I don't fully get yet

File: fixtures/pages-functions-with-routes-app/tests/index.test.ts [hunk 9]
```diff
@@ -53,35 +53,27 @@ describe("Pages Functions with custom _routes.json", () => {
...
-		 *
-		 * `cd fixtures/pages-functions-with-routes-app && npx wrangler pages publish public`
-		 * for now if you want to test that
-		 *
-		 */
 		response = await waitUntilReady("http://localhost:8776/date");
 		text = await response.text();
-		expect(text).toMatch(/\d{4}-\d\d-\d\dT\d\d:\d\d:\d\d/);
+		expect(text).toContain(
+			"Bienvenue sur notre projet &#10024; pages-functions-with-routes-app!"
```
Comment (GregBrimble) [inline fixtures/pages-functions-with-routes-app/tests/index.test.ts:76]: The `/date` route is in `include` and is handled by `_worker.js`. Why does it not return a date here?

File: packages/wrangler/src/pages/dev.tsx [hunk 10]
```diff
@@ -361,7 +466,7 @@ export const Handler = async ({
 			experimentalEnableLocalPersistence,
 			showInteractiveDevSession: undefined,
 			inspect: true,
-			logLevel: "error",
+			logLevel: "warn",
```
Comment (GregBrimble) [inline packages/wrangler/src/pages/dev.tsx:469]: Why did you change this?

Comment (GregBrimble) [reply packages/wrangler/src/pages/dev.tsx:469]: Ah, sorry, I misunderstood what this meant. I thought it was setting all logging to a warning level. I didn't understand that it meant it would _filter_ logs below this level. Gotcha now. Yeah, that's fine 👍

File: packages/wrangler/templates/pages-dev-pipeline.ts (hunk 2 above)
Comment (GregBrimble) [reply packages/wrangler/templates/pages-dev-pipeline.ts:11]: Not easily done here. This template runs in the user-space, whereas that `src/pages` stuff is compiled into wrangler. I think it's good enough as you've got it now :)

File: fixtures/pages-functions-with-routes-app/tests/index.test.ts (hunk 9 above)
Comment (GregBrimble) [reply fixtures/pages-functions-with-routes-app/tests/index.test.ts:76]: Ah, didn't realize that was already in before this PR. I was looking at the wrong file, you're right. All good!

**Review by GregBrimble — APPROVED** [review]
ezgif com-gif-maker

File: packages/wrangler/src/pages/dev.tsx (hunk 10 above)
Comment (CarmenPopoviciu) [reply packages/wrangler/src/pages/dev.tsx:469]: yeah, that's because
```
if (LOGGER_LEVELS[this.loggerLevel] >= LOGGER_LEVELS[messageLevel]) {
	console[messageLevel](this.formatMessage(messageLevel, format(...args)));
}
```
and
```
error: 0,
warn: 1,
log: 2,
debug: 3,
```
